std/math/fp2.asm

```diff
@@ -0,0 +1,147 @@
+use std::convert::fe;
+use std::convert::int;
+use std::convert::expr;
+use std::field::modulus;
+use std::prover::eval;
+
+/// An extension field element, represented as a tuple (a0, a1) of expressions (which evaluate to field elements).
+/// An element of F_{p^2} is a polynomial a0 + a1 * x (mod x^2 - 7).
+/// The polynomial x^2 - 7 is irreducible in both the Goldilocks and BN254 field
+/// (which follows from 7 not having a square root in both fields).
+enum Fp2<T> {
+    Fp2(T, T)
+}
+
+let<T: FromLiteral> zero_ext: -> Fp2<T> = || Fp2::Fp2(0, 0);
+let<T: FromLiteral> one_ext: -> Fp2<T> = || Fp2::Fp2(1, 0);
+
+/// Extension field addition
+let<T: Add> add_ext: Fp2<T>, Fp2<T> -> Fp2<T> = |a, b| match (a, b) {
+    (Fp2::Fp2(a0, a1), Fp2::Fp2(b0, b1)) => Fp2::Fp2(
+        a0 + b0,
+        a1 + b1
+    )
+};
+
+/// Extension field subtraction
+let<T: Sub> sub_ext: Fp2<T>, Fp2<T> -> Fp2<T> = |a, b| match (a, b) {
+    (Fp2::Fp2(a0, a1), Fp2::Fp2(b0, b1)) => Fp2::Fp2(
+        a0 - b0,
+        a1 - b1
+    )
+};
+
+/// Extension field multiplication
+let<T: Add + FromLiteral + Mul> mul_ext: Fp2<T>, Fp2<T> -> Fp2<T> = |a, b| match (a, b) {
+    (Fp2::Fp2(a0, a1), Fp2::Fp2(b0, b1)) => Fp2::Fp2(
+        // Multiplication modulo the polynomial x^2 - 7. We'll use the fact
+        // that x^2 == 7 (mod x^2 - 7), so:
+        // (a0 + a1 * x) * (b0 + b1 * x) = a0 * b0 + 7 * a1 * b1 + (a1 * b0 + a0 * b1) * x (mod x^2 - 7)
+        a0 * b0 + 7 * a1 * b1,
+        a1 * b0 + a0 * b1
+    )
+};
+
+/// Converts and Fp2<expr> into an Fp2<fe>
+let eval_ext: Fp2<expr> -> Fp2<fe> = query |a| match a {
+    Fp2::Fp2(a0, a1) => Fp2::Fp2(eval(a0), eval(a1))
+};
+
+/// Converts and Fp2<fe> into an Fp2<expr>
+let expr_ext: Fp2<fe> -> Fp2<expr> = |a| match a {
+    Fp2::Fp2(a0, a1) => Fp2::Fp2(expr(a0), expr(a1))
+};
+
+/// Extension field equality
```

Comment:
Maybe we should only implement that for fe? Not sure if expr has Eq, but it would be syntactic equality I would assume.